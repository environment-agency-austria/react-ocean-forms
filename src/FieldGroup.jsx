/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';

import { formContextShape, validationShape, fieldValueShape } from './shapes';
import withValidation from './hocs/withValidation';
import FormContext from './FormContext';

/**
 * Wrapper for groups of input fields
 * managed by the form component
 */
class FieldGroup extends React.Component {
  constructor(props) {
    super(props);

    this.validate = this.validate.bind(this);
    this.reset = this.reset.bind(this);

    this.notifyFieldEvent = this.notifyFieldEvent.bind(this);

    const {
      fullName,
      label,
      context,
      validation: {
        update: updateValidation,
      },
    } = props;

    this.checkFormContext();

    // Our state will overwrite the parts of the formContext
    this.state = {
      fieldPrefix: fullName,
      notifyFieldEvent: this.notifyFieldEvent,
    };

    // Register the group in the formContext, so the group
    // validation can be called on form submit.
    context.registerField(
      fullName,
      {
        label,

        validate: this.validate,
        updateValidation,
        reset: this.reset,
        getValue: () => ({}),

        isGroup: true,
      },
    );
  }

  /**
   * Unregisters the field from the form
   */
  componentWillUnmount() {
    const { context, fullName } = this.props;
    context.unregisterField(fullName);
  }

  /**
   * Helper function to get the correct value
   * of the group (including all values of the nested fields)
   */
  getGroupValue() {
    const { context, fullName } = this.props;
    const formValues = context.getValues();
    return formValues[fullName] || '';
  }

  /**
   * Returns the correct asyncValidateOnChange setting,
   * where the field setting takes priorty over the
   * form setting
   */
  getAsyncValidateOnChangeSetting() {
    const {
      asyncValidateOnChange: propChange,
      context: { asyncValidateOnChange: contextChange },
    } = this.props;

    return propChange === null ? contextChange : propChange;
  }

  /**
   * Generates the form context for the FieldGroup children
   * Overrides fieldPrefix and notifyFieldEvent from the
   * parent context, and overrides defaultValues and values
   * for the group items if needed
   */
  getSubContext() {
    const { context } = this.props;

    return {
      ...context,
      ...this.state,
      ...this.overrideContextValues('defaultValues'),
      ...this.overrideContextValues('values'),
    };
  }

  /**
   * Checks if the FieldGroup is inside a valid form context
   * and throws an user friendly error if not
   */
  checkFormContext() {
    const { context, fullName } = this.props;
    if (!context || typeof context.registerField !== 'function') {
      throw new Error(
        `Could not find a form context for field group "${fullName}". `
        + 'Fields can only be used inside a Form tag.',
      );
    }
  }

  /**
   * Checks if the FieldGroup has a prop with the given name
   * and overrides the according value in the parent form context.
   * @param {string} name Property name
   */
  overrideContextValues(name) {
    const {
      fullName,
      context: { [name]: contextValue },
      [name]: propValue,
    } = this.props;

    if (propValue === undefined) {
      return {
        [name]: contextValue,
      };
    }

    return {
      [name]: {
        ...contextValue,
        ...{
          [fullName]: propValue,
        },
      },
    };
  }

  /**
   * Resets the validation state
   */
  reset() {
    const { validation: { reset } } = this.props;
    reset();
  }

  /**
   * Listens to child field events, triggers validation if
   * needed and passes them to the higher context
   * @param {string} name Field name
   * @param {string} event Event name
   * @param {object} args Event args
   */
  notifyFieldEvent(name, event, args) {
    const { fullName, context, validation: { validate } } = this.props;
    context.notifyFieldEvent(name, event, args);

    if (event === 'change' || event === 'blur') {
      const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
      if (event === 'change') {
        const localName = name.substring(fullName.length + 1);

        validate(
          {
            ...this.getGroupValue(),
            ...{
              // Override the value of the event sender, because
              // the Field didn't update its state yet, making the
              // Form.getValues() returning an old Field value.
              [localName]: args,
            },
          },
          { checkAsync: asyncValidateOnChange },
        );
      } else if (!asyncValidateOnChange) {
        validate(this.getGroupValue());
      }
    }
  }

  /**
   * Triggers the validation of the group
   * @param {object} args Options for the validate call
   */
  validate(args) {
    const { validation: { validate } } = this.props;

    // Overwrite the value of the group state with
    // the parsed one.
    const value = this.getGroupValue();
    return validate(value, args);
  }

  render() {
    const {
      fullName,
      validation: {
        isValidating,
        valid,
        error,
      },
      render: renderProp,
    } = this.props;

    const groupState = {
      fullName,
      isValidating,
      valid,
      error,
    };

    const subContext = this.getSubContext();

    return (
      <FormContext.Provider value={subContext}>
        {renderProp(groupState)}
      </FormContext.Provider>
    );
  }
}

FieldGroup.displayName = 'FieldGroup';

FieldGroup.defaultProps = {
  asyncValidateOnChange: null,
  defaultValues: undefined,
  values: undefined,
};

FieldGroup.propTypes = {
  fullName: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  context: formContextShape.isRequired,
  render: PropTypes.func.isRequired,
  validation: validationShape.isRequired,
  asyncValidateOnChange: PropTypes.bool,
  defaultValues: PropTypes.objectOf(fieldValueShape),
  values: PropTypes.objectOf(fieldValueShape),
};

export const BaseFieldGroup = FieldGroup;
export default withValidation(FieldGroup);
