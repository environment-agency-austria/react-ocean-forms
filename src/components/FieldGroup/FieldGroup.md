### Examples

#### Simple field group

Showcase of a simple field group

```jsx
import { Form, FieldGroup, Input } from 'react-ocean-forms';

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = (values) => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  // Notice how a field group creates a sub object
  const defaultValues = {
    myGroup: {
      myInput1: 'input 1',
      myInput2: 'input 2',
    },
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} defaultValues={defaultValues}>
      <FieldGroup
        name="myGroup"
        label="Example group"
        render={() => {
          return (
            <React.Fragment>
              <Input name="myInput1" label="Example input" />
              <Input name="myInput2" label="Example input 2" />
            </React.Fragment>
          );
        }}
      />
      <button type="submit">Submit</button>
    </Form>
  );
}

<Example />;
```
