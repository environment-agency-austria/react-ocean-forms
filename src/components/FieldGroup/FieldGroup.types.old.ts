/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IValidationState } from '../../hooks';
import { IFieldValues } from '../FormContext';
import { IValidationProps } from '../withValidation';

/**
 * Meta information about the group
 * for the render prop
 */
export interface IFieldGroupRenderParams extends IValidationState {
  /**
   * Full name of the group
   */
  fullName: string;
}

/**
 * Props for the field group component
 */
export interface IFieldGroupProps<TFieldValues = IFieldValues> extends IValidationProps<TFieldValues> {
  /**
   * Field name
   */
  name: string;
  /**
   * Label of the group
   */
  label: string;
  /**
   * True, if the async validators should be triggered
   * on a change event
   */
  asyncValidateOnChange?: boolean;
  /**
   * Optional default values
   */
  defaultValues?: TFieldValues;
  /**
   * Optional values
   */
  values?: TFieldValues;
  /**
   * Disables this field group and all its fields.
   */
  disabled?: boolean;
  /**
   * Puts the field group and all its fields in plaintext mode.
   */
  plaintext?: boolean;
  /**
   * Render prop
   * @param params Meta information about the group
   */
  render(params: IFieldGroupRenderParams): JSX.Element;
}
