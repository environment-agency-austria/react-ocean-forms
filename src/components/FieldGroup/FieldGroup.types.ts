/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IUseFieldGroupArgs, IFieldGroupRenderParams } from './hooks/useFieldGroup.types';

/**
 * Props for the field group component
 */
export interface IFieldGroupProps extends IUseFieldGroupArgs {
  /**
   * Gets called to render its children (see render prop pattern).
   * @param params Contains the group state consisting of fullName,
   * isValidating,valid, error which can be used to display those informations.
   */
  render(params: IFieldGroupRenderParams): JSX.Element;
}
