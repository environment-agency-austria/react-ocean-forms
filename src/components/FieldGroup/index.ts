export * from './FieldGroup';
export * from './FieldGroup.types';
export { IFieldGroupRenderParams } from './hooks/useFieldGroup.types';
