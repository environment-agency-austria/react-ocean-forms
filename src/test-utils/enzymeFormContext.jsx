import { stringFormatter } from '../utils';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
export const createMockFormContext = registerCallback => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name, state) => registerCallback(name, state)),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: jest.fn().mockImplementation(stringFormatter),

  busy: false,
  disabled: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a validation object mocking the
 * withValidation hoc
 */
export const createMockValidation = () => ({
  validate: jest.fn(),
  reset: jest.fn(),
  update: jest.fn(),

  isValidating: false,
  valid: true,
  error: null,
});
