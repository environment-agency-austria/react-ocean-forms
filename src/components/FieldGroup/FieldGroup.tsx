/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { FormContext } from '../FormContext';
import { IFieldGroupProps } from './FieldGroup.types';
import { useFieldGroup } from './hooks/useFieldGroup';

/**
 * Wrapper for groups of input fields
 * managed by the form component
 */
export const FieldGroup: React.FC<IFieldGroupProps> = ({
  name, validators, asyncValidators, asyncValidationWait, render,
  disabled, plaintext, label, asyncValidateOnChange,
}): JSX.Element => {
  const [ subContext, groupState ] = useFieldGroup(
    name, label, validators, asyncValidators, asyncValidationWait,
    disabled, plaintext, asyncValidateOnChange,
  );

  return (
    <FormContext.Provider value={subContext}>
      {render(groupState)}
    </FormContext.Provider>
  );
}
