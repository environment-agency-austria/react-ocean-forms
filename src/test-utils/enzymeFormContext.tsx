import { IFieldState, IFormContext, TFormEventListener } from '../components/FormContext';
import { IValidationProp } from '../components/withValidation';
import { stringFormatter } from '../utils';

/**
 * Components inside the form module require access to the form context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * user definable form context around them.
 */

/**
 * Creates a form context
 */
// tslint:disable-next-line:naming-convention
export const createMockFormContext = (registerCallback?: Function): IFormContext => ({
  fieldPrefix: null,

  registerField: jest.fn().mockImplementation((name: string, state: IFieldState): void => {
    if (registerCallback) {
      registerCallback(name, state);
    }
  }),
  unregisterField: jest.fn(),
  notifyFieldEvent: jest.fn(),

  registerListener: jest.fn().mockImplementation((name: string, callback: TFormEventListener): void => {
    if (registerCallback) {
      registerCallback(name, callback);
    }
  }),
  unregisterListener: jest.fn(),

  getFieldState: jest.fn(),
  getValues: jest.fn(),

  stringFormatter: jest.fn().mockImplementation(stringFormatter),
  submit: jest.fn(),

  busy: false,
  disabled: false,
  plaintext: false,

  asyncValidateOnChange: false,
  asyncValidationWait: 400,
  defaultValues: {},
});

/**
 * Creates a validation object mocking the
 * withValidation hoc
 */
// tslint:disable-next-line:naming-convention
export const createMockValidation = (): IValidationProp => ({
  validate: jest.fn(),
  reset: jest.fn(),
  update: jest.fn(),

  isValidating: false,
  isRequired: false,
  valid: true,
  error: null,
});
