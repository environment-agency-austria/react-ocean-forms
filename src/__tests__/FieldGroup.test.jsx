import React from 'react';
import { shallow } from 'enzyme';

import { createMockFormContext, createMockValidation } from '../test-utils/enzymeFormContext';
import { BaseFieldGroup } from '../FieldGroup';

describe('<FieldGroup />', () => {
  let fieldState = null;
  const registerCallback = (name, state) => { fieldState = state; };
  const renderMock = jest.fn().mockReturnValue(null);

  const fieldName = 'unitGroup';
  const fullName = fieldName;
  const fieldLabel = 'Unit group';
  const validation = createMockValidation();
  const formContext = createMockFormContext(registerCallback);

  const mockGroupValue = {
    field1: 'value1',
    field2: 'value2',
  };
  formContext.getValues = jest.fn().mockReturnValue({
    foo: 'bar',
    hue: 'hue',
    [fieldName]: mockGroupValue,
  });

  const setup = props => shallow((
    <BaseFieldGroup
      name={fieldName}
      fullName={fullName}
      label={fieldLabel}
      context={formContext}
      validation={validation}
      render={renderMock}
      {...props}
    />
  ));
  const wrapper = setup();

  it('should render without error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should register itself in the form context', () => {
    expect(formContext.registerField).toHaveBeenCalledWith(
      fieldName,
      {
        label: fieldLabel,
        validate: expect.any(Function),
        reset: expect.any(Function),
        getValue: expect.any(Function),
        updateValidation: validation.update,
        isGroup: true,
      },
    );
  });

  describe('notifyFieldEvent proxy behaviour', () => {
    const eventSender = `${fieldName}.demoName`;
    const changeValue = 'blubb';
    const groupValue = {
      demoName: changeValue,
      ...mockGroupValue,
    };

    const getNotifyCallback = () => wrapper.instance().notifyFieldEvent;
    const triggerNotification = (event) => {
      const notify = getNotifyCallback();
      notify(eventSender, event, changeValue);
    };

    it('should pass any event through to the form context', () => {
      const notify = getNotifyCallback();

      const eventName = 'blubb';
      const eventArgs = { foo: 'bar' };

      notify(eventSender, eventName, eventArgs);
      expect(formContext.notifyFieldEvent).toHaveBeenCalledWith(eventSender, eventName, eventArgs);
    });

    it('should call the validate function onChange', () => {
      triggerNotification('change');

      expect(validation.validate).toHaveBeenLastCalledWith(
        groupValue,
        { checkAsync: false },
      );
    });

    it('should respect the Form.asyncValidateOnChange configuration', () => {
      formContext.asyncValidateOnChange = true;
      wrapper.setProps({ context: formContext });

      triggerNotification('change');
      expect(validation.validate).toHaveBeenLastCalledWith(
        groupValue,
        { checkAsync: true },
      );
    });

    it('should respect the Field.asyncValidateOnChange configuration', () => {
      formContext.asyncValidateOnChange = false;
      wrapper.setProps({ context: formContext, asyncValidateOnChange: true });

      triggerNotification('change');
      expect(validation.validate).toHaveBeenLastCalledWith(
        groupValue,
        { checkAsync: true },
      );

      wrapper.setProps({ asyncValidateOnChange: false });
    });

    it('should validate onBlur', () => {
      triggerNotification('blur');
      expect(validation.validate).toHaveBeenLastCalledWith(groupValue);
    });
  });

  describe('form context callbacks', () => {
    it('should return an empty object as the value', () => {
      expect(fieldState.getValue()).toMatchObject({});
    });

    it('should correctly pass the validate method', () => {
      const validateArgs = { foo: 'bar' };

      fieldState.validate(validateArgs);
      expect(validation.validate).toHaveBeenLastCalledWith(mockGroupValue, validateArgs);
    });

    it('should correctly reset to its defaultValue', () => {
      fieldState.reset();
      wrapper.update();

      expect(validation.reset).toHaveBeenCalled();
    });
  });

  it('should correctly unregister on unmount', () => {
    wrapper.unmount();
    expect(formContext.unregisterField).toHaveBeenCalledWith(fieldName);
  });
});
