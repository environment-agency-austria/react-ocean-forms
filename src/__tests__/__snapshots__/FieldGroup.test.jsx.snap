// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldGroup /> should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FieldGroup
    asyncValidateOnChange={null}
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "registerField": [MockFunction] {
          "calls": Array [
            Array [
              "unitGroup",
              Object {
                "getValue": [Function],
                "isGroup": true,
                "label": "Unit group",
                "reset": [Function],
                "updateValidation": [MockFunction],
                "validate": [Function],
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    fullName="unitGroup"
    label="Unit group"
    name="unitGroup"
    render={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "error": null,
              "fullName": "unitGroup",
              "isValidating": false,
              "valid": true,
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": null,
          },
        ],
      }
    }
    validation={
      Object {
        "error": null,
        "isValidating": false,
        "reset": [MockFunction],
        "update": [MockFunction],
        "valid": true,
        "validate": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": null,
      "value": Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": "unitGroup",
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [Function],
        "registerField": [MockFunction] {
          "calls": Array [
            Array [
              "unitGroup",
              Object {
                "getValue": [Function],
                "isGroup": true,
                "label": "Unit group",
                "reset": [Function],
                "updateValidation": [MockFunction],
                "validate": [Function],
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      },
    },
    "ref": null,
    "rendered": null,
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": "form",
        "_currentValue2": "form",
        "_defaultValue": "form",
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": null,
        "value": Object {
          "asyncValidateOnChange": false,
          "asyncValidationWait": 400,
          "busy": false,
          "defaultValues": Object {},
          "disabled": false,
          "fieldPrefix": "unitGroup",
          "getFieldState": [MockFunction],
          "getValues": [MockFunction],
          "notifyFieldEvent": [Function],
          "registerField": [MockFunction] {
            "calls": Array [
              Array [
                "unitGroup",
                Object {
                  "getValue": [Function],
                  "isGroup": true,
                  "label": "Unit group",
                  "reset": [Function],
                  "updateValidation": [MockFunction],
                  "validate": [Function],
                },
              ],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          },
          "registerListener": [MockFunction],
          "stringFormatter": [MockFunction],
          "unregisterField": [MockFunction],
          "unregisterListener": [MockFunction],
        },
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": "form",
          "_currentValue2": "form",
          "_defaultValue": "form",
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
